src/infra/core/src/repos/metadata_chain_visitor_impl.rs

```diff
@@ -0,0 +1,340 @@
+        }
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+pub struct ValidatePrevBlockExistsVisitor {
+    next_prev_block_hash: Option<Multihash>,
+}
+
+impl MetadataChainVisitor for ValidatePrevBlockExistsVisitor {
+    fn visit(
+        &mut self,
+        (block, maybe_hash): MetadataBlockWithOptionalHashRef,
+    ) -> Result<Decision, AppendError> {
+        let Some(next_prev_block_hash) = &self.next_prev_block_hash else {
+            self.next_prev_block_hash = block.prev_block_hash.clone();
+
+            return Ok(Decision::Next);
+        };
+
+        if next_prev_block_hash != maybe_hash.unwrap() {
+            return Err(
+                AppendValidationError::PrevBlockNotFound(BlockNotFoundError {
+                    hash: next_prev_block_hash.clone(),
+                })
+                .into(),
+            );
+        }
+
+        Ok(Decision::Stop)
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+pub struct ValidateSequenceNumbersIntegrityVisitor {
+    next_sequence_number: Option<u64>,
+}
+
+impl ValidateSequenceNumbersIntegrityVisitor {
+    fn process_first_block(
+        &mut self,
+        first_block: &MetadataBlock,
+    ) -> Result<Decision, AppendError> {
+        self.next_sequence_number = Some(first_block.sequence_number);
+
+        if first_block.prev_block_hash.is_none() && first_block.sequence_number != 0 {
+            return Err(
+                AppendValidationError::SequenceIntegrity(SequenceIntegrityError {
+                    prev_block_hash: None,
+                    prev_block_sequence_number: None,
+                    next_block_sequence_number: first_block.sequence_number,
+                })
+                .into(),
+            );
+        }
+
+        Ok(Decision::Next)
+    }
+}
+
+impl MetadataChainVisitor for ValidateSequenceNumbersIntegrityVisitor {
+    fn visit(
+        &mut self,
+        (block, maybe_hash): MetadataBlockWithOptionalHashRef,
+    ) -> Result<Decision, AppendError> {
+        let Some(next_sequence_number) = self.next_sequence_number else {
+            return self.process_first_block(block);
+        };
+
+        if block.sequence_number != (next_sequence_number - 1) {
+            return Err(
+                AppendValidationError::SequenceIntegrity(SequenceIntegrityError {
+                    prev_block_hash: Some(maybe_hash.unwrap().clone()),
+                    prev_block_sequence_number: Some(block.sequence_number),
+                    next_block_sequence_number: next_sequence_number,
+                })
+                .into(),
+            );
+        }
+
+        Ok(Decision::Stop)
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+pub struct ValidateSystemTimeIsMonotonicVisitor {
+    next_system_time: Option<DateTime<Utc>>,
```

Comment:
Looks like many visitors have to start with some empty state and then initialize it with the first block...

I wonder if instead of:
```rust
self.accept_with_extra_block(block, [
  VisitorA::default(),
  VisitorB::default(),
])
```
it would be more ergonomic to use them as:
```rust
self.accept([
  VisitorA::new(&new_block),
  VisitorB::new(&new_block),
])
```
allowing visitors to initialize their state from first block and eliminate `Option`s.

Reply:
This would eliminate the need for a special `accept` method that passes an extra block.

Reply:
API has been updated:
```rust
pub trait MetadataChainVisitor: Sync + Send {
    type VisitError: Error;

    fn visit(
        &mut self,
        hashed_block_ref: HashedMetadataBlockRef,
    ) -> Result<Decision, Self::VisitError>;
}
```